### PR TITLE
Support ReadableStream in request.clone & response.clone()

### LINKED ISSUE #1381: `Request.clone()` does not clone or support streaming body
### What version of Bun is running?

0.2.1 (canary)

### What platform is your computer?

Darwin 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 20:57:23 PDT 2022; root:xnu-8020.121.3~2/RELEASE_ARM64_T8110 arm64 arm

### What steps can reproduce the bug?

```ts
export default {
  async fetch(request) {
    console.log(".clone().json()", await request.clone().json());
    console.log(".json()", await request.json());
    return new Response("Ok");
  }
}
```

`curl -d '{}' http://localhost:3000`

### How often does it reproduce? Is there a required condition?

Always. Same with `.text()`.

### What is the expected behavior?

_No response_

### What do you see instead?

Terminal output:
```bash
.clone().json() # body is missing
.json() {} # works fine
```

### Additional information

_No response_

### LINKED ISSUE #2477: `request.clone()` -> `Request.prototype.json()` not working from HTTP server
Reported: https://discord.com/channels/876711213126520882/876711213126520885/1088916040387395614

The following snippet:
```js
Bun.serve({
  async fetch(request) {
    const req = request.clone();
    const b = await req.json();
    const a = await request.json();
    return new Response(`Bun!`);
  },
});
```

gives me this error with Bun v0.5.8 (and it worked with v0.5.7)
```
16 | Bun.serve({
17 |   async fetch(request) {
18 |     const req = request.clone();
                    ^
SyntaxError: Unexpected end of JSON input
```

### LINKED ISSUE #2644: FormData missing final boundary
### What version of Bun is running?

0.5.9

### What platform is your computer?

WSL - x64

### What steps can reproduce the bug?

```js
async function handler(req: Request) {
	const contentType = req.headers.get("Content-Type")
	if (!contentType) throw new Error("Missing Content-Type");

	if (contentType.startsWith("multipart/form-data")) {
		const fd = await req.formData()
		const body = Object.fromEntries(fd.entries());
		return new Response(JSON.stringify(body))
	}

	throw new Error("Invalid Content-Type");
}
Bun.serve({
	port: 3000,
	async fetch(request) {
		await new Promise((resolve) => setTimeout(resolve, 0));
		const response = await handler(request)
		await new Promise((resolve) => setTimeout(resolve, 0));
		return response;
	}
})
```
Here is the code causing the bug. Try to hit `localhost:3000` many times quickly and the server will throw the following error message: 
`error: FormData missing final boundary`

When removing the setTimeouts everything works correctly. That is what causing the problem.

### What is the expected behavior?

Not to throw an error

### What do you see instead?

The error described above.

### Additional information

_No response_

### LINKED ISSUE #4161: TypeError: FormData parse error missing final boundary
### What version of Bun is running?

0.7.3

### What platform is your computer?

Linux 5.15.0-60-generic x86_64 x86_64

### What steps can reproduce the bug?

I'm using Bun.serve` with the `request.formData` 

I'm using sendgrid to receive emails in bun but Sendgrid is apparently not giving FormData dat you can parse.
Somehow, sometimes it works, sometimes it doesn't.

Before I go more in-depth... maybe you can look at your `request.formData()` or let me know how to use it?

Makes it very difficult to get the files of attachments of emails in sendgrid.

### What is the expected behavior?

It should parse it even without boundary

### What do you see instead?

It throws `TypeError: FormData parse error missing final boundary`

### Additional information

My code:

```
export const tryGetFormData = async (request: Request | undefined) => {
  try {
    return request?.clone().formData();
  } catch (e) {
    return undefined;
  }
};
```

It Throws with Sendgrid incoming mail

### LINKED ISSUE #6348: Response.clone does not clone in-progress streaming body
### What version of Bun is running?

1.0.4

### What platform is your computer?

Linux 5.10.60.1-microsoft-standard-WSL2 x86_64 x86_64

### What steps can reproduce the bug?

[Calling the `clone()` method on a `Response` object should return an identical `Response` object, including the body](https://developer.mozilla.org/en-US/docs/Web/API/Response/clone). Instead, it returns a `Response` that's missing the body, but otherwise identical in every other way (that I've found).

Minimal reproducible example:
```js
const res = await fetch("https://github.com/");
console.log("Clone Response length:", (await res.clone().text()).length)
console.log("Main Response length:", (await res.text()).length)
```

### What is the expected behavior?

The program should output two lines to the console, and each should contain the same integer number.

### What do you see instead?

The first line will contain the number "0" while the second line contains the expected value.

### Additional information

_No response_

### LINKED ISSUE #10789: TypeError: FormData parse error missing final boundary in SolidStart
### What version of Bun is running?

1.1.6+e58d67b46

### What platform is your computer?

Linux 6.5.0-26-generic x86_64 x86_64

### What steps can reproduce the bug?

https://github.com/huseeiin/bun-sld-pwd
you have to submit any of the forms in the index page to see the error.

### What is the expected behavior?

for the FormData to be parsed correctly.

### What do you see instead?

image

### Additional information

**important detail**: it works in prod for some reason
```bash
bun --bun run build
bun .output/server/index.mjs
```

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

Fixes #6348
Fixes #1381
Fixes #2477
Fixes #2644
Fixes #10789
Fixes #4161

### How did you verify your code works?

Updated the body stream tests.

Still need to update some more tests. But let's see what CI says.

### CODE REVIEW

**Comment by robobun** [comment]
@Jarred-Sumner, your commit 191e22926141f983c665d104b3547b8b6722796d is building: [`#2787`](https://buildkite.com/bun/bun/builds/2787)

File: src/bun.js/webcore/response.zig [hunk 1]
```diff
@@ -261,10 +261,27 @@ pub const Response = struct {
...
+            return .zero;
+        }
+
+        const js_wrapper = Response.makeMaybePooled(globalThis, cloned);
+
+        if (js_wrapper != .zero) {
+            if (cloned.body.value == .Locked) {
+                if (cloned.body.value.Locked.readable.get()) |readable| {
+                    Response.bodySetCached(js_wrapper, globalThis, readable.value);
+                    Response.bodySetCached(this_value, globalThis, this.body.value.Locked.readable.get().?.value);
```
Comment (cirospaciari) [inline src/bun.js/webcore/response.zig]: ✂️
```suggestion
                    
```

File: src/codegen/bundle-functions.ts [hunk 2]
```diff
@@ -148,12 +160,23 @@ async function processFileSplit(filename: string): Promise<{ functions: BundledB
         globalThis.requireTransformer(x, SRC_DIR + "/" + basename),
       );
 
+      const source = result.trim().slice(2, -1);
+      const enums: string[] = [];
+      if (topLevelEnums.length) {
+        for (const { name, code } of topLevelEnums) {
+          if (source.includes(name)) {
+            enums.push(code);
+          }
```
Comment (cirospaciari) [inline src/codegen/bundle-functions.ts:174]: can you add some comment on this?

File: src/bun.js/webcore/response.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bun.js/webcore/response.zig]: This is intentional. tee() returns 2 readable streams

File: src/codegen/bundle-functions.ts (hunk 2 above)
Comment (Jarred-Sumner) [reply src/codegen/bundle-functions.ts:174]: yeah

**Review by dylan-conway — APPROVED** [review]

File: src/bun.js/webcore/response.zig (hunk 1 above)
Comment (cirospaciari) [reply src/bun.js/webcore/response.zig]: I miss read sorry

**Review by cirospaciari — APPROVED** [review]

**Pull request merged**